pkg/aws/policies.go

```diff
@@ -0,0 +1,215 @@
+	// (i.e. ec2:StartInstances, iam:ChangePassword)
+	Action []string `json:"Action"`
+	// If you create an IAM permissions policy, you must specify a list of resources to which
+	// the actions apply. If you create a resource-based policy, this element is optional. If
+	// you do not include this element, then the resource to which the action applies is the
+	// resource to which the policy is attached.
+	Resource []string `json:"Resource"`
+}
+
+type PolicyStatementPrincipal struct {
+	// A service principal is an identifier that is used to grant permissions to a service.
+	// The identifier for a service principal includes the service name, and is usually in the
+	// following format: service-name.amazonaws.com
+	Service []string `json:"Service"`
+	// You can specify an individual IAM role ARN (or array of role ARNs) as the principal.
+	// In IAM roles, the Principal element in the role's trust policy specifies who can assume the role.
+	// When you specify more than one principal in the element, you grant permissions to each principal.
+	AWS []string `json:"AWS"`
+}
+
+func (c *awsClient) EnsureRole(name string, policy string, tagList map[string]string) (string, error) {
+	output, err := c.iamClient.GetRole(&iam.GetRoleInput{
+		RoleName: aws.String(name),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeNoSuchEntityException:
+				return c.createRole(name, policy, tagList)
+			default:
+				return "", err
+			}
+		}
+	}
+	role := output.Role
+	return aws.StringValue(role.Arn), nil
+}
+
+func (c *awsClient) createRole(name string, policy string, tagList map[string]string) (string, error) {
+	output, err := c.iamClient.CreateRole(&iam.CreateRoleInput{
+		RoleName:                 aws.String(name),
+		AssumeRolePolicyDocument: aws.String(policy),
+		Tags:                     getTags(tagList),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException:
+				return "", nil
+			}
+		}
+		return "", err
+	}
+	role := output.Role
+	return aws.StringValue(role.Arn), nil
+}
+
+func (c *awsClient) PutRolePolicy(roleName string, policyName string, policy string) error {
+	_, err := c.iamClient.PutRolePolicy(&iam.PutRolePolicyInput{
+		RoleName:       aws.String(roleName),
+		PolicyName:     aws.String(policyName),
+		PolicyDocument: aws.String(policy),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *awsClient) EnsurePolicy(name string, document string, tagList map[string]string) error {
+	_, err := c.iamClient.CreatePolicy(&iam.CreatePolicyInput{
+		PolicyName:     aws.String(name),
+		PolicyDocument: aws.String(document),
+		Tags:           getTags(tagList),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException:
+				return nil
+			}
+		}
+		return err
+	}
```

Comment:
Is it a conscious decision to not return an error if the policy already exists? How would this work when a customer upgrades from `4.7` to `4.8` for instance? Applying the `4.8` policies over the `4.7` policies for instance results in no change even though the policy for `4.8` adds a permission `elasticloadbalancing:DeregisterTargets` which is silently dropped in this case.

Reply:
Good question. I'll add a way to not just ensure the role exists, but make sure it's the right version. Maybe I can add it in a follow-up PR?

Reply:
My concern is we push this release and we're close to the 4.8 upgrade. It should be easy enough to validate the `rosa_openshift_version` tag and return an error if it doesn't match? If its a bigger change then ok let's do it in a separate PR.

Reply:
@vkareh lets merge this one and call out the risk before we cut a release.